### PR TITLE
Fix quantum channel delay

### LINKED ISSUE #528: Bell pairs generation time does not seem to follow theoretical calculation
**Describe the bug**

When running a link simulation, entanglement generation without purification or entanglement swap, the time to generate Bell pairs is quicker than the theoretical values. This means that either the loss probability is incorrect or the delay in the communication channel is set incorrectly. 

**To Reproduce**

The theoretical time to reach a certain number of Bell pairs for MIM link can be described by
```python
def calculate_time(L, R, N_pair, N_memory):
    print(f"L: {L}, R: {R}")
    p_bsa = 0.5
    p_loss = 0.0450074
    c = 208189
    t_attemp = 2 * max(L, R) / c
    p_entangle = (1 - p_loss) ** (L+R) * p_bsa
    T = N_pair * t_attemp / p_entangle / N_memory
    return T
```
But the numbers seen from QuISP simulation are less than the theoretical ones.

### PR DESCRIPTION
**Pull request by Naphann**
This PR fixes the behaviour that the Bell pairs are generated too fast, due to the quantum channel having no delay.

The changes are made in this PR:
- introduce QuantumChannel delay from the distance which used to be overwritten from `processMessage` function
- fix the incorrect unit conversion for the indistinguishable time window for photon entanglement at BSA

This closes #528.

- - -
This change is [Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/529)

### CODE REVIEW

**Review by res6idue — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/529)* status: all files reviewed, 1 unresolved discussion (waiting on @Chibikuri, @Naphann, and @zigen)

___
*[`quisp/modules/PhysicalConnection/BSA/BellStateAnalyzer.cc` line 28 at r1](https://reviewable.io/reviews/sfc-aqua/quisp/529#-NZKKCvS-lPerUibKuRb:-NZKKCvS-lPerUibKuRc:bq4qtp0) ([raw file](https://github.com/sfc-aqua/quisp/blob/66ded62e5a38180e11cdbcf2f9575255968ed950/quisp/modules/PhysicalConnection/BSA/BellStateAnalyzer.cc#L28)):*
> ```cpp
>   darkcount_probability = par("darkcount_probability").doubleValue();
>   detection_efficiency = par("detection_efficiency").doubleValue();
>   indistinguishability_window = SimTime(par("indistinguishable_time_window").doubleValue() * 1000, SIMTIME_PS);
> ```

I don't know if this is my misunderstanding, but even if we have this code and the quantum channel is wrong, we seem to obtain bell pairs successfully.

**Review by res6idue — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/529)* status: all files reviewed, 1 unresolved discussion (waiting on @Chibikuri, @Naphann, and @zigen)

___
*[`quisp/modules/PhysicalConnection/BSA/BellStateAnalyzer.cc` line 28 at r1](https://reviewable.io/reviews/sfc-aqua/quisp/529#-NZKKCvS-lPerUibKuRb:-NZKLC_B-WDgRuaoNCsg:b-5skbok) ([raw file](https://github.com/sfc-aqua/quisp/blob/66ded62e5a38180e11cdbcf2f9575255968ed950/quisp/modules/PhysicalConnection/BSA/BellStateAnalyzer.cc#L28)):*
<details><summary><i>Previously, res6idue (soon) wrote…</i></summary><blockquote>

I don't know if this is my misunderstanding, but even if we have this code and the quantum channel is wrong, we seem to obtain bell pairs successfully. 
</blockquote></details>

Also, out of curiosity, why is this doing *10^3 in PS? Why can't we use NS?

**Review by Naphann — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/529)* status: all files reviewed, 1 unresolved discussion (waiting on @Chibikuri, @res6idue, and @zigen)

___
*[`quisp/modules/PhysicalConnection/BSA/BellStateAnalyzer.cc` line 28 at r1](https://reviewable.io/reviews/sfc-aqua/quisp/529#-NZKKCvS-lPerUibKuRb:-NZKLr6q6ScS12_el6WG:b-m2a81h) ([raw file](https://github.com/sfc-aqua/quisp/blob/66ded62e5a38180e11cdbcf2f9575255968ed950/quisp/modules/PhysicalConnection/BSA/BellStateAnalyzer.cc#L28)):*
<details><summary><i>Previously, res6idue (soon) wrote…</i></summary><blockquote>

Also, out of curiosity, why is this doing *10^3 in PS? Why can't we use NS?
</blockquote></details>

I don't understand the the first comment. 

What we want is to represent the input time window in the correct unit. The old code reads the double value and translate it to whatever the resolution was set. In the usual case it will be on the order of milliseconds which will be too long. Meaning that even if one of the photon arrive 100microsecond later the BSA still consider a successful measurement.

By doing this, you can see test that if the BSA is 1km away from Alice and 19km away from Bob and while the Channel has no delay. You will not get any Bell pairs.

As for the second question, the constructor accepts a long long, so the default 1.5ns (which is a floating point) will get round down to 1ns. So I use PS instead to truncate on the order of PS.

**Review by res6idue — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/529)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Chibikuri and @zigen)

___
*[`quisp/modules/PhysicalConnection/BSA/BellStateAnalyzer.cc` line 28 at r1](https://reviewable.io/reviews/sfc-aqua/quisp/529#-NZKKCvS-lPerUibKuRb:-NZKNhYh75dSJbpVa9PF:b14qd3q) ([raw file](https://github.com/sfc-aqua/quisp/blob/66ded62e5a38180e11cdbcf2f9575255968ed950/quisp/modules/PhysicalConnection/BSA/BellStateAnalyzer.cc#L28)):*
<details><summary><i>Previously, Naphann (Naphan Benchasattabuse) wrote…</i></summary><blockquote>

I don't understand the the first comment. 

What we want is to represent the input time window in the correct unit. The old code reads the double value and translate it to whatever the resolution was set. In the usual case it will be on the order of milliseconds which will be too long. Meaning that even if one of the photon arrive 100microsecond later the BSA still consider a successful measurement.

By doing this, you can see test that if the BSA is 1km away from Alice and 19km away from Bob and while the Channel has no delay. You will not get any Bell pairs.

As for the second question, the constructor accepts a long long, so the default 1.5ns (which is a floating point) will get round down to 1ns. So I use PS instead to truncate on the order of PS. 
</blockquote></details>

For the first comment, I would apologize that I only tested the code on equidistant nodes...
I have just checked it for 1-19km unbalanced nodes on my side as well, and now it is working as intended.

**Review by res6idue — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/529#-:-NZKOA9Z-n5snD1TNsMA:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Chibikuri and @zigen)

**Review by zigen — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/529#-:-NZOWVI8AyGSruqF968q:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Chibikuri)

**Review by Chibikuri — APPROVED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/529)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Naphann)

**Review by Chibikuri — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/sfc-aqua/quisp/529#-:-NZO_S-18u23zuPnVm3I:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Naphann)
